### PR TITLE
feat(reconciliation): Do hybrid reconciliation

### PR DESCRIPTION
**Pull request by fredrb**
## Description

This is a PR attempting to fix the reconciliation by still keeping deletes in central, but sending a set of `unchangedIDs` to central at the `SYNC` message, as a batch `SYNC` for any resource that hasn't changed. 

A more detailed description of the problem and solutions can be found [here](https://docs.google.com/document/d/1m9xgPAg9-Q_ukcAKxbHH11_oKl-__qQCb2CJ_NvMGKc/edit?usp=sharing).

## Checklist
- [x] Investigated and inspected CI test results
- [x] Unit test and regression tests added
- [ ] ~~Evaluated and added CHANGELOG entry if required~~
- [ ] ~~Determined and documented upgrade steps~~
- [ ] ~~Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~~

If any of these don't apply, please comment below.

### Here I tell how I validated my change

Turned the feature flag on, and let CI run with new reconciliation. Given that this PR makes use of Central reconciliation, all reconciliation related tests shall serve as a regression test.

- [x] CI 
    - [x] ReconciliationTest
    - [x] UpgradeTest (backup recover)

### Reminder for reviewers

In addition to reviewing code here, reviewers **must** also review testing and request further testing in case the
performed one does not seem sufficient. As a reviewer, you must not approve the change until you understand the
performed testing and you are satisfied with it.

### CODE REVIEW

**Comment by vikin91** [comment]
@fredrb is this ready for a review? If yes, then please fill the `Here I tell how I validated my change`, as I would like to follow the idea for solution while looking at the code.

File: sensor/common/sensor/central_sender_impl.go
Comment (vikin91) [inline sensor/common/sensor/central_sender_impl.go:117]: By looking at this diff, the `onSync` function seems not to be used anymore. Could we remove it maybe?

**Review by vikin91 — COMMENTED** [review]
The PR looks really well! Thanks for explaining me the details of this idea on Slack and in the doc. 

I am leaving few comments to the code but none of them are blockers.
What bothers me a bit are the :x: from the CI. I would like to wait with reviewing ✅ until they are resolved.

**Comment by fredrb** [comment]
> What bothers me a bit are the ❌ from the CI

@vikin91 for some reason `s390x` build failed on my PR, but I couldn't figure out why. This caused images to not be pushed and thus prow being entirely red. 

I have rebased with master and fixed your comments. If this build fails again I'll investigate further.

File: sensor/common/deduper/unchanged.go
Comment (vikin91) [inline sensor/common/deduper/unchanged.go:10]:
```suggestion
// ClosableSet is a set of strings. Items can be added to the set as long as the set is open. Closing the set ignores all future `Add` operations but allows to fetch already stored items. 
```

File: sensor/common/deduper/unchanged.go
Comment (vikin91) [inline sensor/common/deduper/unchanged.go:17]: Update comment pls.

File: sensor/common/deduper/unchanged.go
Comment (vikin91) [inline sensor/common/deduper/unchanged.go:28]:
```suggestion
// AddIfOpen adds item to an unclosed set. If set is closed, this operation does nothing
```

File: sensor/common/deduper/unchanged.go
Comment (vikin91) [inline sensor/common/deduper/unchanged.go:37]:
```suggestion
// Close marks set as closed preventing any new items to be added. It returns the currently stored items.
```

**Review by vikin91 — COMMENTED** [review]
I think, the doc-comments to the ClosableSet could be refreshed a bit.

**Review by vikin91 — APPROVED** [review]
Thanks, it looks good! The IBM build is green now, so let's wait for the CI.
You can also polish the comments for the `ClosableSet` if there is an opportunity for pushing another commit.

File: sensor/common/deduper/unchanged.go
Comment (fredrb) [reply sensor/common/deduper/unchanged.go]: Done. Just changed to also add info about it being concurrency-safe.

File: sensor/common/deduper/deduper.go
Comment (lvalerom) [inline sensor/common/deduper/deduper.go:91]: [nit] This functions are also defined [here](https://github.com/stackrox/stackrox/blob/4ba873b685b734595366b78880c6c2af8ada9e9d/central/hash/manager/deduper.go#L124-L131). Would it make sense to maybe extract them in a common package under `pkg`?

**Review by lvalerom — APPROVED** [review]
Looks good! Sorry for the late review 😅

**Comment by fredrb** [comment]
- `gke-ui-e2e-tests` seem to be failing for every PR.
- `gke-version-compatibility-tests` failed on the post test but all tests seem to have passed. I will re-run them just in case.

File: sensor/common/deduper/deduper.go
Comment (fredrb) [reply sensor/common/deduper/deduper.go:91]: Created a follow-up for this: [ROX-20693](https://issues.redhat.com/browse/ROX-20693)
I think there might be other things we can pull up to a shared package.

File: central/sensor/service/connection/sensorevents.go
Comment (connorgorman) [inline central/sensor/service/connection/sensorevents.go:100]: How much can this message be realistically?

File: pkg/deduperkey/key.go
Comment (connorgorman) [inline pkg/deduperkey/key.go:50]: This check isn't necessary so I think it's better to omit it

File: proto/internalapi/central/sensor_events.proto
Comment (connorgorman) [inline proto/internalapi/central/sensor_events.proto:67]: My only concern is about the size of this message

File: sensor/common/deduper/unchanged.go
Comment (connorgorman) [inline sensor/common/deduper/unchanged.go:12]: I'm confused by this a bit. Why would you ever add anything to a closed set? Doesn't that imply a timing bug where a synced resource is trying to be added after close?

File: sensor/common/deduper/unchanged.go
Comment (connorgorman) [inline sensor/common/deduper/unchanged.go:31]: nit: I find it cleaner to check the opposite cases and then return which avoids extra indentation

File: sensor/debugger/central/communicate_server.go
Comment (connorgorman) [inline sensor/debugger/central/communicate_server.go:82]: No need to initialize a mutex

File: sensor/common/sensor/central_sender_impl.go
Comment (connorgorman) [inline sensor/common/sensor/central_sender_impl.go:115]: This is a strange usage of the getters. It's cleaner if you modify this in the if above as in

```
if syncEvent := msg.GetEvent().GetSynced(); syncEvent != nil 
```

then you operate on syncEvent

File: sensor/common/sensor/central_sender_impl.go
Comment (connorgorman) [inline sensor/common/sensor/central_sender_impl.go:120]: This section seems unnecessary? just inc by the len(of unchanged ids) and print out the ideas using %+v?

File: sensor/common/deduper/deduper.go
Comment (connorgorman) [inline sensor/common/deduper/deduper.go:66]: This is a mega nit so feel free to ignore, but I think it should be Closeable

**Review by connorgorman — COMMENTED** [review]
I added some comments, most of which will not impact the code though please respond to some of the conversations before merging

File: sensor/common/deduper/unchanged.go
Comment (fredrb) [reply sensor/common/deduper/unchanged.go]: >  Why would you ever add anything to a closed set?

We shouldn't ever. That's why the set is being closed after the reconciliation finishes.

> Doesn't that imply a timing bug where a synced resource is trying to be added after close?

It does, because the set should be closed as soon as sensor observes a `ResourcesSynced` event. If any sensor event gets triggered with `ResourceAction_SYNC_RESOURCE` it's a timing issue.

But my decision was to ignore that event rather than log an error. I will change this a bit to show an error if this case happens.

File: sensor/common/deduper/unchanged.go
Comment (fredrb) [reply sensor/common/deduper/unchanged.go]: Done

File: sensor/common/deduper/deduper.go
Comment (fredrb) [reply sensor/common/deduper/deduper.go]: Done

File: central/sensor/service/connection/sensorevents.go
Comment (fredrb) [reply central/sensor/service/connection/sensorevents.go:100]: note: fallback to classic reconciliation if max message receive is received.

File: pkg/features/list.go
Comment (fredrb) [inline pkg/features/list.go:68]: Switch to false

File: pkg/deduperkey/key.go
Comment (fredrb) [reply pkg/deduperkey/key.go:57]: Done

File: sensor/common/deduper/unchanged.go
Comment (fredrb) [reply sensor/common/deduper/unchanged.go]: I've changed the code and ended up removing this struct all together. I think it's simpler now.

File: sensor/debugger/central/communicate_server.go
Comment (fredrb) [reply sensor/debugger/central/communicate_server.go:82]: Done

File: sensor/common/sensor/central_sender_impl.go
Comment (fredrb) [reply sensor/common/sensor/central_sender_impl.go:115]: Done

File: sensor/common/sensor/central_sender_impl.go
Comment (fredrb) [reply sensor/common/sensor/central_sender_impl.go:120]: Removed

File: central/sensor/service/connection/sensorevents.go
Comment (fredrb) [reply central/sensor/service/connection/sensorevents.go:100]: Created a task to address the fallback to classic reconciliation [ROX-20878](https://issues.redhat.com/browse/ROX-20878).

File: proto/internalapi/central/sensor_events.proto
Comment (fredrb) [reply proto/internalapi/central/sensor_events.proto:67]: See comment above about problem doc and the new task to implement classic reconciliation fallback.

**Pull request merged**
